dataLayer Builder + Enhanced Ecommerce.tpl

```diff
@@ -0,0 +1,554 @@
+                  },
+                  {
+                    "type": 8,
+                    "boolean": false
+                  }
+                ]
+              }
+            ]
+          }
+        }
+      ]
+    },
+    "clientAnnotations": {
+      "isEditedByUser": true
+    },
+    "isRequired": true
+  }
+]
+
+
+___SANDBOXED_JS_FOR_WEB_TEMPLATE___
+
+const makeTableMap = require('makeTableMap');
+const query = require('queryPermission');
+const createQueue = require('createQueue');
+const dataLayerPush = createQueue('dataLayer');
+const log = require('logToConsole');
+const merge = function() {
+  const obj = {},
+        il = arguments.length;
+  let i = 0,
+      key;
+  for (; i < il; i++) {
+    for (key in arguments[i]) {
+      if (arguments[i].hasOwnProperty(key)) {
+        obj[key] = arguments[i][key];
+      }
+    }
+  }
+  return obj;
+};
+const event = {'event': data.event};
+const currencyParameter = data.currencyParameter;
+let actionFieldParameters = data.actionFieldParameters ? makeTableMap(data.actionFieldParameters, 'name', 'value') : {};
+const actionFieldDropDown = data.actionFieldDropDown;
+ if (actionFieldDropDown === true) {
+    actionFieldParameters = actionFieldParameters;
+ } else {
+   actionFieldParameters = undefined;
+ }
+
+const products = data.products;
+let customParameter = data.customParameter ? makeTableMap(data.customParameter, 'name', 'value') : {};
+const customParameterDropDown = data.customParameterDropDown;
+
+if (customParameterDropDown === true) {
+  customParameter = customParameter;
+} else {
+  customParameter = undefined;
+}
+//const customParameter = data.customParameter ? makeTableMap(data.customParameter, 'name', 'value') : {};
+const eventType = data.eventType;
+var dLayer = '';
+if (eventType == 'impressions') {
+  dLayer = {
+  'ecommerce': {
+    'currencyCode': currencyParameter,
+    'actionField': actionFieldParameters,
+    'impressions': products 
+ }
+};
+} else if (eventType == 'click') {
+ dLayer = {
+  'ecommerce': {
+    'currencyCode': currencyParameter,
+    'click': { 
+     'actionField': actionFieldParameters,
+     'products': products
+  } 
+ }
+}; 
+} else if (eventType == 'detail') {
+ dLayer = {
+  'ecommerce': {
+    'currencyCode': currencyParameter,
+    'detail': {
+     'actionField': actionFieldParameters, 
+     'products':  products
+  }
+ }
+};
+} else if (eventType == 'add') {
+ dLayer = {
+  'ecommerce': {
+    'currencyCode': currencyParameter,
+    'add': { 
+     'actionField': actionFieldParameters, 
+     'products':  products
+  }
+ }
+};
+} else if (eventType == 'remove') {
+ dLayer = {
+  'ecommerce': {
+    'currencyCode': currencyParameter,
+    'remove': { 
+     'actionField': actionFieldParameters, 
+     'products':  products 
+  }
+ }
+}; 
+} else if (eventType == 'promoView') {
+ dLayer = {
+  'ecommerce': {
+    'currencyCode': currencyParameter,
+    'promoView': { 
+     'actionField': actionFieldParameters, 
```

Comment:
No `actionField` in promoView either :)

Reply:
I have removed the actionField parameter from the '**promoClick**' object in the custom template. I do however have an observation and question. I can't seem to set a 2nd enabling condition in the template to block for both '**Impressions**' and '**promoClick**' since there is no option to use RebEx, and it appears you can't have 2 enabling conditions in a row as they appear to cancel each other out.
I did try to create my own variable **const actionFieldEnable = ((eventType == ('impressions' || 'promoClick'))) ? 'false' : 'true';** and use this variable in the enabling conditions to block or fire based on a boolean but it didn't work either. It seems like custom templates is a bit limited without the use of RegEx or supporting the use of your own custom variables to enable fields. Do you have any suggestions on how to block enabling on multiple events?
 
Thanks in advance!

Reply:
Yeah, you can't use a "block" condition due to the thing you described (I've given feedback about this). 

You need to turn it around, and add all the "equals" options as possibilities.

So instead of someVar does not equal "one" and someVar does not equal "two", you have

someVar equals three
someVar equals four
someVar equals five

That will do until we get the regex.

Reply:
Thanks for the helpful tip, this did the trick, albeit very inefficiently! :) I really am hoping that Google will add in support for the RegEx object in the future, this will make all our lives a lot easier. 

Also, I did some very minor field name updates to make them a bit clearer. If you could do one more once over that would be great. This template should be good to go! 

Once more thing, apologies for not thinking about overwriting the main project readme, in future, I will make sure to put the whole think its own folder so this doesn't happen with future builds. 

Thank!